### PR TITLE
Introduce a Representor Namespace

### PR DESCRIPTION
**Pull request by kylef**
This pull request is a work in progress pull request to add a namespace for the Representor pattern as described in the [hypermedia charter](https://github.com/the-hypermedia-project/charter).

I'm not entirely sure this is what we're looking for. Can someone (@zdne?) take a look at point in in what direction I should be taking this.

### CODE REVIEW

**Comment by fosrias** [comment]
@kylef So, I think the more fundamental question here is not so much the representor pattern in refract vs. the canonical resource description that is a subset of the canonical api description language we are talking about. The actual representor interface is an opinionated interface that will wrap a refract object that is the source of truth on the canonical representation. Not sure if this is making sense, and if not, maybe we can jump on a chat. There is a lot of stuff @smizell and I have been discussing around refract, our current and future tooling architecture that we have not had a chance to vet with you and @kylef. Maybe we need to set up a meeting to do that early next week. Thoughts?

I can certainly chat with you a little about this outside of a group meeting as well.

**Comment by smizell** [comment]
Additionally, I think Verbose will end up being a place to find the semantics for this.

**Comment by zdne** [comment]
> There is a lot of stuff @smizell and I have been discussing around refract

@fosrias @smizell seems like you guys took over this whole thing – we should discuss the situation and future work. I feel like albeit on our list most of the work was done by Stephen and not Blueprint team. We need to decide how to proceed with representor and API description namespaces.

File: namespaces/Representor.md [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+This document extends [Refract](../refract-spec.md) Specification to describe the [Representor](https://github.com/the-hypermedia-project/charter) DOM.
+
+## Representor (Object Element)
+
+The Representor represents a Hypermedia Resource, with it's available links, transitions and attributes.
+
+### Properties
+
+- `links` - (Array Element of Link) - The links from this resource.
```
Comment (zdne) [inline namespaces/Representor.md]: `(array[Link])` is the MSON for array of "Links"

Comment (zdne) [inline namespaces/Representor.md]: Also I am afraid, the `Link` in this namespace is clashing with `Link` from the base spec. Sure we can namespace it but maybe it would be easier to pick another name

File: namespaces/Representor.md [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+- `transitions` - (Array Element of Transition) - The available transitions from this resource.
+- `representors` - (Array Element of Representor) - The embedded links (resources) from this resource.
+- `attributes` - (Object Element) - Any properties of the resource.
```
Comment (zdne) [inline namespaces/Representor.md]: So this would be MSON here, correct?

File: namespaces/Representor.md [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+- `metadata` - (Object Element) _ Meta Attributes, for example, may include a `title` and `description`.
+
+## Link (Object Element)
```
Comment (zdne) [inline namespaces/Representor.md]: Mind clash with https://github.com/refractproject/refract-spec/blob/master/refract-spec.md#link-enum

File: namespaces/Representor.md [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+A link element representing a link and it's relation from the Representor.
+
+### Properties
+
+- `relation` - (String Element, required) - The relation from the Representor to this link.
```
Comment (zdne) [inline namespaces/Representor.md]: Also shouldn't this be just `string` instead of element?

File: namespaces/Representor.md (hunk 1 above)
Comment (zdne) [inline namespaces/Representor.md]: I think you want to follow Refract here. So instead top-level properties, nest this under either `content` or `attributes`. The way you have written it here does not follow refract concepts where everything is an element of `element`, `meta`, `attributes` and `content` properties.

File: namespaces/Representor.md [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+- `uri` - (String Element, required) - The URI for this link.
+
+### Example
+
+```json
+{
+    "element": "object",
+    "content": [
```
Comment (zdne) [inline namespaces/Representor.md]: Oh I see it now, but this isn't quite what is written in MSON on line 23 (you are missing the `content` property there

File: namespaces/Representor.md [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+## Transition (Object Element)
+
+A transition is an available progression from one state to another state.
+
+### Properties
+
+- `relation` - (String Element, required) - The relation of this transition from the current resource.
+- `uri` - (String Element, required) - The URI for this transition.
+- `parameters` (Array Element of Input Property)
+- `attributes` (Array Element of Input Property)
```
Comment (zdne) [inline namespaces/Representor.md]: It is questionable whether these properties are in `content` or `attributes` property of `transition` element.

**Comment by zdne** [comment]
@kylef I think you have missed how Refract elements are defined:

```
# New Element (Element)
- element: `new-element-type-name` (string, fixed) 
- attributes - any new element specific attributes here
- content - any new element specific content here 
```

**Comment by zdne** [comment]
@kylef  Re 

> I'm not entirely sure this is what we're looking for.

We basically need a sets of elements that enables us to capture a resource description (e.g. from blueprint) in a refract DOM (think of it as refract equivalent of https://github.com/apiaryio/api-blueprint-ast#resource-object)

**Comment by kylef** [comment]
@zdne Yes, I understand now. I've updated to address this feedback.

File: namespaces/Representor.md [hunk 7]
```diff
@@ -0,0 +1,141 @@
+# Representor Namespace
+
+This document extends [Refract](../refract-spec.md) Specification to describe the [Representor](https://github.com/the-hypermedia-project/charter) DOM.
+
+## Resource (Object Element)
+
+The Resource represents a Hypermedia Resource, with it's available relations, transitions and attributes.
```
Comment (zdne) [inline namespaces/Representor.md]: its (?)

File: namespaces/Representor.md [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+### Properties
+
+- `element`: resource (string, fixed)
+- `content`: (object)
```
Comment (zdne) [inline namespaces/Representor.md]: Note, in MSON the `:` is used to denote a sample value e.g. `content: hello!`, it probably shouldn't be here

File: namespaces/Representor.md [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+    - `relations` - (array[Relation]) - The relations from this resource.
```
Comment (zdne) [inline namespaces/Representor.md]: What is the purpose of `relations` if you have `transitions`? I believe `transition` is listing all the possible transition including their relations?

File: namespaces/Representor.md [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+    - `transitions` - (array[Transition]) - The available transitions from this resource.
+    - `resources` - (array[Resource]) - The embedded relations (resources) from this resource.
+    - `attributes` - (object) - Any properties of the resource.
```
Comment (zdne) [inline namespaces/Representor.md]: If I understand this right – this is the actual "content" of the resource and therefore it should be the "content" of the `content` property (see bellow).

File: namespaces/Representor.md [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+    - `relation` - (string) - An optional relation to this resource.
```
Comment (zdne) [inline namespaces/Representor.md]: I am completely missing this one – what is it good for? I assume it says what "relation" was exercised to get to this resource. Is this used anywhere in the representor pattern?

File: namespaces/Representor.md [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+
```
Comment (zdne) [inline namespaces/Representor.md]: Focusing on what is the real content of a resource vs. some metadata – what about this:

## Resource (Object Element)
- `element`: resource (string, fixed)
- `attributes` 
  - `transitions` - (array[Transition]) - The available transitions from this resource
  - `resources` - (array[Resource]) - The embedded relations – resources – from this resource
- `content` (MSON Element) - The content of the Element
  
  Either as Refract primitives (imagine a JSON here) or as MSON elements. 

### Example

``` json
{
    "element": "resource",
    "attributes": {
        "transitions": [
            {
                "relation": "self",
                "uri": "/user"
            }
        ]
    },
    "content": {
        "name": "John",
        "address": "john@apple.com"
    }
}
```

The idea there is that if you take the content you will get the actual resource data (content). Either directly or as an MSON

File: namespaces/Representor.md (hunk 11 above)
Comment (zdne) [inline namespaces/Representor.md]: Also a typo in "optional" however I tend to not mention this in description as it should be implied by `optional` or `required` attribute.

**Comment by zdne** [comment]
@kylef I think this is going in the right direction. From this proposal is clear that we need to introduce 4 elements here `resource`, `transition`, `relation` and `inputProperty`.

However I believe we should utilize the concept of data (`content`) and metadata (`attributes`) to take advantage of simple serialization (e.g. into JSON). 

Also I am missing any references to MSON namespaces which – I believe – is the actual building stone for the `resource` element.

File: namespaces/Representor.md (hunk 11 above)
Comment (kylef) [inline namespaces/Representor.md]: This is the relation of this resource (useful when it's referenced from another resource).

I.e:

``` json
{
    "element": "resource",
    "resources": [
        {
            "element": "resource",
            "content": {
                "relation": "choice"
            }
        }
    ]
}
```

**Comment by kylef** [comment]
@zdne Just the feedback I was looking for. I think initially I mis-understood that the attributes could be extended and wasn't reserved for Refract itself. I've updated the pull request to make these changes. I'm not sure what to do with `Input Property` so that is left as-is, would love any ideas anyone has with this one.

File: namespaces/Representor.md [hunk 13]
```diff
@@ -0,0 +1,107 @@
+# Representor Namespace
+
+This document extends [Refract](../refract-spec.md) Specification to describe the [Representor](https://github.com/the-hypermedia-project/charter) DOM.
+
+## Resource (Object Element)
```
Comment (zdne) [inline namespaces/Representor.md]: I think `Element` instead of `Object Element` is sufficient here, because you are overriding the rest anyway. Or am I missing something?

File: namespaces/Representor.md [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+
+The Resource represents a Hypermedia Resource, with its available relations, transitions and attributes.
+
+### Properties
+
+- `element`: resource (string, fixed)
+- `attributes` (object)
+    - `relations` - (array[Relation]) - The relations from this resource.
+    - `transitions` - (array[Transition]) - The available transitions from this resource.
```
Comment (zdne) [inline namespaces/Representor.md:13]: I feel like I have asked this already but why both `relations` and `transitions` – isn't it possible to get all `relations` from the `transitions`?

File: namespaces/Representor.md [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+    - `resources` - (array[Resource]) - The embedded relations (resources) from this resource.
+    - `relation` - (string) - An optional relation to this resource, when embedded in another resource.
+- `content` (object) - Any properties of the resource.
```
Comment (zdne) [inline namespaces/Representor.md:16]: I think the `content` shouldn't be `object` but `MSON Element`. 

``` json
["resource", {}, {}, ["array", {}, {}, null]]

["resource", {}, {}, ["object", {}, {}, null]]
```

etc.

File: namespaces/Representor.md [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+
+## Relation (Object Element)
+
+A relation element representing a relation to another resource.
```
Comment (zdne) [inline namespaces/Representor.md]: Relation really blurs with with transition. Is this the "Siren" approach (or my ignorance of hypermedia project' charter). Do we define `GET` transitions as "relation" whereas anything else is just "transition"? cc @fosrias @smizell 

In other words: **What is the difference between transition and relation?**

File: namespaces/Representor.md [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+
+### Properties
+
+- `element`: relation (string, fixed)
+- `attributes` (object, required)
+    - `relation` - (string, required) - The relation from the Resource to this resource.
+- `content` (string, required) - The URI for this relation.
```
Comment (zdne) [inline namespaces/Representor.md]: I wonder whether we need to address relative vs absolute URLs here..

File: namespaces/Representor.md [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+
+- `element`: resource (string, fixed)
+- `attributes` (object)
+    - `relations` - (array[Relation]) - The relations from this resource.
+    - `transitions` - (array[Transition]) - The available transitions from this resource.
+    - `resources` - (array[Resource]) - The embedded relations (resources) from this resource.
+    - `relation` - (string) - An optional relation to this resource, when embedded in another resource.
+- `content` (object) - Any properties of the resource.
+
+## Relation (Object Element)
```
Comment (zdne) [inline namespaces/Representor.md]: `Element` should do

File: namespaces/Representor.md [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+
+## Transition (Object Element)
+
+A transition is an available progression from one state to another state.
+
+### Properties
+
+- `element`: transition (string, fixed)
+- `attributes` (object, required)
+    - `relation` - (string, required) - The relation of this transition from the current resource.
```
Comment (zdne) [inline namespaces/Representor.md:26]: I found it odd that sometimes `relation` is `string` and sometimes it means `Relation (object)`. Is this the "relation type"? But I guess it all goes down to my relation vs transition question.

File: namespaces/Representor.md [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+```
+
+## Input Property
+
+An Input Property is used to describe a potential property or attribute used to perform a transition.
+
+### Properties
+
+- `element`: property (string, fixed)
+- `content`: (object)
```
Comment (zdne) [inline namespaces/Representor.md]: Shouldn't this be simply `MSON Element` (and the rest is not needed)?

File: namespaces/Representor.md [hunk 21]
```diff
@@ -0,0 +1,107 @@
...
+}
+```
+
+## Input Property
+
+An Input Property is used to describe a potential property or attribute used to perform a transition.
+
+### Properties
+
+- `element`: property (string, fixed)
```
Comment (zdne) [inline namespaces/Representor.md]: also I would probably go with `inputProperty` to prevent any ambiguity with object properties...

**Comment by zdne** [comment]
I think this is almost good to go. I would like a clarification on relation (objet) vs transition (object) vs relation (string).

**Comment by zdne** [comment]
@kylef @fosrias let's have a quick chat about this so we can close it...

**Comment by zdne** [comment]
@kylef also I think this namespace should contain an "HTTP layer" 

I was thinking about these lines:

``` json
["httpTransaction", {}, { "requestMethod": "GET" }, [ 
    ["transactionExample", {}, {}, [
        ["request", {}, {}, [
            "payload", {}, {"headers": {}, "attributes": null}, null
        ]]
        ["response", {}, {"status": 200}, [
            "payload", {}, {"headers": {},  "attributes": null}, null
        ]]        
    ]],
    ["transactionExample", {}, {}, []] 
]]
```

But "payload" element probably needs some tweaks when it comes to HTTP headers, content-types and incorporating MSON namespace attributes description...

**Comment by zdne** [comment]
Putting refract aside the transition should be probably able to bear all the information specified in API Blueprint action – https://github.com/apiaryio/api-blueprint-ast#action-object

File: namespaces/Representor.md [hunk 22]
```diff
@@ -0,0 +1,44 @@
...
+
+A transition is an available progression from one state to another state.
+
+### Properties
+
+- `element`: transition (string, fixed)
+- `attributes` (object, required)
+    - `relation` - (string, required) - The relation of this transition from the current resource.
+    - `parameters` ([Object Type][]) - A collection of parameters for this transition.
+    - `attributes` ([Object Type][]) - A collection of possible attributes for this transition.
```
Comment (smizell) [inline namespaces/Representor.md:28]: I have a few questions/comments here.

First, why the MSON Object Type instead of the Refract Object Element? I have felt for a while we've been putting MSON things into the base Refract spec, and even brought up whether the entirety of MSON should be in the base spec. Using the MSON Object Type here makes it seem to me like something is off, either the way we've defined these specs or the way this is being used.

Second, I am not sure about mixing primitive JSON objects with Refract elements like this. Notice [here](https://github.com/refractproject/refract-spec/blob/master/refract-spec.md#properties) how the `attributes` property is an `enum` of a primitive object and an array of Member Elements. This means that using `attributes` as on object is simply a shortcut for writing out all of these member elements you would normally use. Here, though, you have a normal object with MSON objects mixed in. This does not seem correct to me.

Third, you have `attributes` nested within `attributes`, and it seems like we are missing something here. Additionaly, the content of this Transition element is the URI, and I'm wondering if that is the place where the transition attributes should be. This would mean you'd need an `href` property here for the URI. 

Hope that all makes sense.

Comment (smizell) [inline namespaces/Representor.md:28]: Also, I'm not completely sure about `parameters` either in the `attributes`. 

Really sorry to pick this apart like this, but it feels like we're missing something.

**Comment by smizell** [comment]
I have a few general comments to make about this. 

First, it seems like we need to decide if this namespace is for the API Description elements or for a canonical hypermedia message. I usually consider the latter when I'm thinking of representors, so I'm not sure putting an HTTP layer per @zdne comments above feels right. In other words, this namespace feels like it should be a subset of the API Description namespace.

Second, I have found the easiest way to think of these elements is to think of them as XML first. This seems odd at first, but it's a big help.

In light of this, this is what I had in mind for representor elements when I was thinking through this with Verbose.

``` json
["verbose", {}, {}, 
  ["link", {}, { "rel": "item", "href": "http://example.com/user/1" }],
  ["link", {}, { "rel": "item", "href": "http://example.com/user/2" }],
  ["form", {}, { "rel": "create-user", "href": "http://example.com/user/2", "method": "POST" },
    ["field", {}, { "name": "name", "value": ""}, null]
    ["field", {}, { "name": "email", "value": ""}, null]
  ]
]
```

I think you're going to want to end up with something close to this, even if you combine `link` and `form` I have above in transition. I think it's better to differentiate link and form myself just because you don't have to introspect anything to know a `form` element has fields. That may be a discussion for a different day. 

Third, if an HTTP layer is still being considered, I think you can simplify some of the examples above.

``` json
["transactions", {}, {}, 
  ["transaction", {}, { "method": "POST", "request-uri": "http://example.com/users", "protocol": "HTTP/1.1" }, 
    ["request", {}, { "content-type": "application/json" }, {
      "name": "John Doe",
      "email": "john@example.com"
    }],
    ["response", {}, { "status": "201", "content-type": "application/json"}, {
      "id": 334,
      "name": "John Doe",
      "email": "john@example.com"
    }]
  ]
]
```

The idea is to apply Refract by asking, what is the "thing," what are the attributes of this "thing," and what other "things" does this "thing" relate to? In the above, we treat a request as a "thing" and the HTTP headers as the attributes of that request, with the body being represented by the Refract content. You could make it more robust by using `ref`.

``` json
["transactions", {}, {}, 
  ["transaction", {}, { "method": "POST", "request-uri": "http://example.com/users", "protocol": "HTTP/1.1" }, 
    ["request", {}, { "content-type": "application/json" }, {
      "name": "John Doe",
      "email": "john@example.com"
    }],
    ["response", {}, { "status": "201", "content-type": "application/json"}, 
      ["ref", {}, {}, "some-mson-data-type-id"]
    ]
  ]
]
```

Again, I think these elements fit outside the scope of this representor namespace.

---

Just saying this again, sorry for picking this apart. I see some areas where we are going back to thinking of things as JSON objects, and I mention all of this in hopes of starting some discussion around that.

File: namespaces/Representor.md [hunk 23]
```diff
@@ -0,0 +1,44 @@
...
+## Resource (Element)
+
+The Resource represents a Hypermedia Resource, with its available transitions and attributes.
+
+### Properties
+
+- `element`: resource (string, fixed)
+- `attributes` (object)
+    - `transitions` - (array[Transition]) - The available transitions from this resource.
+    - `resources` - (array[Resource]) - Any embedded (resources) from this resource.
```
Comment (smizell) [inline namespaces/Representor.md:14]: I missed this as well, but embedding other resource elements in the `attributes` does not seem right. It seems like they should be in the content. Would be worth discussing I think.

**Comment by zdne** [comment]
@smizell 

But isn't 

```
  ["form", {}, { "rel": "create-user", "href": "http://example.com/user/2", "method": "POST" },
```

coupled with HTTP due to the "method" field?

**Comment by smizell** [comment]
@zdne The other option is to create some kind of mapping like [UBER does](https://github.com/mamund/media-types/blob/master/uber-hypermedia.asciidoc#33-reserved-strings).

But I also don't see a whole lot of issue in designing this form as a form for instructions on how to build an HTTP request. Maybe there is a good middle ground as this mapping adds a layer of abstraction that may be premature at this point (not sure).

**Comment by smizell** [comment]
Also, you could consider this an HTTP namespace. We could have forms for other application protocols when there arose the need for them, but it doesn't seem like we are interested right now in accommodating them. We are mostly interested in accommodating the ability to use them if desired, which the namespaces in Refract already allow for.

**Comment by zdne** [comment]
@smizell well, let's say if would accommodate the separation from protocol implementation concept of [Resource Blueprint](https://github.com/resource-blueprint/resource-blueprint) I would be happy.

**Comment by smizell** [comment]
@zdne It may be that then we'll need a mapping from canonical protocol semantics and protocol-specific methods like POST or GET to make this happen.

It seems like you'll want a canonical way to map between all of these application protocols. This means that we'll have a canonical API Description format, which includes a canonical way of describing RESTful application protocols and canonical hypermedia messages. This means the sections at the bottom of [this example](https://github.com/resource-blueprint/resource-blueprint/blob/master/examples/Gist%20API.md) could be autogenerated from the canonical definition.

In other words, this would be a singular, canonical vocabulary for the entirety of the application along with mappings (i.e. transformation instructions) to each specific protocol and hypermedia format.

cc: @fosrias

**Comment by smizell** [comment]
> This means the sections at the bottom of this example could be autogenerated from the canonical definition

When I say "sections at the bottom," I mean sections like [HTTP](https://github.com/resource-blueprint/resource-blueprint/blob/master/examples/Gist%20API.md#http) for protocol and the [media types](https://github.com/resource-blueprint/resource-blueprint/blob/master/examples/Gist%20API.md#media-types) section.

**Comment by kylef** [comment]
I've created a pull-request against this pull request to add a place to put protocol information in a generic way: https://github.com/kylef/refract-spec/pull/1

**Comment by zdne** [comment]
@smizell 

> This means the sections at the bottom of this example could be autogenerated from the canonical definition.

Not quite true. In some cases yes, but as with everything, if you want non-generator-default behavior you should be able to provide custom HTTP details.

**Comment by zdne** [comment]
> I've created a pull-request against this pull request to add a place to put protocol information in a generic way

:+1:

**Comment by smizell** [comment]
> Not quite true. In some cases yes, but as with everything, if you want non-generator-default behavior you should be able to provide custom HTTP details.

Ah, makes sense.

File: namespaces/Representor.md (hunk 22 above)
Comment (zdne) [inline namespaces/Representor.md:28]: Note `attributes` can be an array too – so instead of `Object Type` I would use `MSON Element` or just `Element`.

Comment (zdne) [inline namespaces/Representor.md:28]: @smizell 

> Second, I am not sure about mixing primitive JSON objects with Refract elements like this. Notice here how the attributes property is an enum of a primitive object and an array of Member Elements. This means that using attributes as on object is simply a shortcut for writing out all of these member elements you would normally use. Here, though, you have a normal object with MSON objects mixed in. This does not seem correct to me.

I actually do not like this policy "its either plain hash OR array of elements". It actually feels like it is complicating the things (re Jirka's confusion). 

If you are going to constrain it even more in a sense "if it is a plain hash the values CANNOT be elements" I would vote for dropping the "plain hash" completely. :(

Comment (zdne) [inline namespaces/Representor.md:28]: > Third, you have attributes nested within attributes, and it seems like we are missing something here

Are you saying that an elements's attribute cannot be called `attributes`? Note the attribute `attributes` is completely unrelated to refract, rather it represents input properties – a body attribute.

Comment (zdne) [inline namespaces/Representor.md:28]: > First, why the MSON Object Type instead of the Refract Object Element? I have felt for a while we've been putting MSON things into the base Refract spec, and even brought up whether the entirety of MSON should be in the base spec. Using the MSON Object Type here makes it seem to me like something is off, either the way we've defined these specs or the way this is being used.

You MAY need a description of data structure here. That is why. Not just plain data. If we put MSON element here it MAY be just plain data (Element) or data description (MSON Element)

Comment (zdne) [inline namespaces/Representor.md:28]: Thinking about this even more – if you do not want elements in plain attributes hash – _maybe_ the href should be attribute and content of the transition SHOULD be input properties – parameters and attributes elements.

But hey @smizell look at MSON namespace – I am having elements in attributes object too ! – https://github.com/refractproject/refract-spec/blob/master/namespaces/mson-namespace.md#properties (actually there is a typo there and _T_ should be _MSON Element_ :stuck_out_tongue: )

File: namespaces/Representor.md (hunk 23 above)
Comment (zdne) [inline namespaces/Representor.md:14]: > they should be in the content

I would say embedded resources can be omitted when serializing a resource. The actual content of a resource is NOT any embedded resource but the resource data itself. To me it make much sense to have embedded resources in attributes. 

Maybe the whole problem of why this doesn't feel right is the use of word "attributes" together with mixing "plain hash of plain key: values" vs. "everything is element"

File: namespaces/Representor.md (hunk 22 above)
Comment (smizell) [inline namespaces/Representor.md:28]: > Thinking about this even more – if you do not want elements in plain attributes hash – maybe the href should be attribute and content of the transition SHOULD be input properties – parameters and attributes elements.

it's not that it's what I want (I'm no dictator here!), it just feels like it's not designed correctly. I'm not against this practice of using elements like this, but I think it actually complicates things. The attributes hash is a shortcut for a fully-refracted object (array of member elements). You can do this refracting infinitely as a refracted member element itself has attributes. What happens when you refract attributes that contain already-refracted elements?

How you have described it here is how I have it in my mind. Links (IMO) look very similar to HTML `<link>` or `<a>` tags. Forms are the same. The benefit we have is that we can include MSON into a "form" and provide a JSON-like input rather than key/value inputs like with HTML `<inputs>`.

Just to circle back around. Refract is suppose to free you from complex JSON structures to use simple, single-purpose elements. Overloading attributes is going back to the old ways (which again, is not evil :) ).

> You MAY need a description of data structure here.

Could you use the description meta property for this description?

File: namespaces/Representor.md (hunk 23 above)
Comment (smizell) [inline namespaces/Representor.md:14]: The reason why I think embedded resources go into the content is because embedded resources are actually part of the resource representation (though not part of the actual state server-side). This is actually probably a topic with lots of opinions, but if you embed a resource in the message, you are changing the message representation. A linked resource versus an embedded resource, though the same in theory, are actually different practically speaking.

It seems to me that you have two options that make sense:
1. Include it in content as said above with a special "embed" element
2. Have a "link" element that can contain the entirety of the linked resource, making it a link with some embedded data

With all of that said, I'm not against it being in attributes, though it is not really an attribute of the resource to me and (again) feels like an anti-pattern.

**Comment by zdne** [comment]
To be addressed in #13
